KAFKA-3751: SASL/SCRAM implementation

rajinisivaram opened the pull request:
Implementation of KIP-84: https://cwiki.apache.org/confluence/display/KAFKA/KIP-84%3A+Support+SASL+SCRAM+mechanisms

rajinisivaram committed:
KAFKA-3751: SASL/SCRAM implementation

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramCredentialCache.java:
```diff
@@ -0,0 +1,60 @@
+        return cacheMap.get(mechanism);
+    }
+
+    public class Cache {
```
Should this be a static class?

rajinisivaram replied:
Yes, changed.

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramFormatter.java:
```diff
@@ -0,0 +1,174 @@
+
+/**
+ * Scram message salt and hash functions defined in <a href="https://tools.ietf.org/html/rfc5802">RFC 5802</a>.
+ *
```
Nit: unnecessary new line?

rajinisivaram replied:
Removed.

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramFormatter.java:
```diff
@@ -0,0 +1,174 @@
+        return hash(clientKey);
+    }
+
+    public String saslname(String username) {
```
`saslName`?

rajinisivaram replied:
Done.

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramMechanism.java:
```diff
@@ -0,0 +1,84 @@
+        String scramPrefix = "SCRAM-";
+        if (!mechanismName.startsWith(scramPrefix))
+            throw new IllegalArgumentException("Invalid mechanism " + mechanismName);
+        this.hashAlgorithm = mechanismName.substring(scramPrefix.length());
```
Nit: instead of passing the full string and then doing a substring on it, I think the other way around is a little bit clearer. That is, pass the components of the mechanism name separately and join them in the constructor.

rajinisivaram replied:
Done.

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramMechanism.java:
```diff
@@ -0,0 +1,84 @@
+    private final String hashAlgorithm;
+    private final String macAlgorithm;
+
+    private static final String[] ALL_MECHANISMS;
```
Do we use this anywhere? Maybe we don't need a field for it?

rajinisivaram replied:
Removed.

ijuma commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -96,6 +102,26 @@ object ConfigCommand extends Config {
+    def scramCredential(mechanism: ScramMechanism, credentialStr: String): String = {
+      val pattern = "(?:iterations=([0-9]*),)?password=(.*)".r
+      val (iterations, password) =
+        credentialStr match {
```
Nit: this can be in the previous line.

rajinisivaram replied:
Done.

ijuma commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -277,7 +304,7 @@ object ConfigCommand extends Config {
             "For entity_type '" + ConfigType.Topic + "': " + LogConfig.configNames.map("\t" + _).mkString(nl, nl, nl) +
             "For entity_type '" + ConfigType.Broker + "': " + DynamicConfig.Broker.names.asScala.map("\t" + _).mkString(nl, nl, nl) +
-            "For entity_type '" + ConfigType.User + "': " + DynamicConfig.Client.names.asScala.map("\t" + _).mkString(nl, nl, nl) +
+            "For entity_type '" + ConfigType.User + "': " + DynamicConfig.User.names.asScala.map("\t" + _).mkString(nl, nl, nl) +
```
Good catch.

ijuma commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -49,7 +50,7 @@ import scala.util.control.{ControlThrowable, NonFatal}
  *   M Handler threads that handle requests and produce responses back to the processor threads for writing.
  */
-class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time) extends Logging with KafkaMetricsGroup {
+class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time, val credentialCache: Option[ScramCredentialCache] = None) extends Logging with KafkaMetricsGroup {
```
It's a bit odd to have a `ScramCredentialsCache` in `SocketServer` and other generic classes (like `ConfigHandler`. It would be nice if we could abstract it so that these classes don't know about specific authentication mechanisms.

rajinisivaram replied:
I have made the cache more generic since delegation tokens will also require a cache, but with credentials of a different type. Have moved authentication related code into a separate class.

ijuma reviewed (commented):
Sorry for the delay and thanks for the PR. I did a first pass to get a feel for the changes and left mostly minor comments. The one important comment is the one regarding how generic classes should ideally not know about specific authentication mechanisms. It would be good to figure out if there's a better way.

In terms of getting this merged, I think there are 2 parts:

1. Changes in the code to accommodate SCRAM. This is relatively straightforward.
2. The SCRAM implementation code. This will require a bit more time as I will have to understand the specification in enough detail to validate the implementation.

rajinisivaram committed:
KAFKA-3751: Address review comments

rajinisivaram commented:
@ijuma Thank you for the review. Have made the changes suggested.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/CredentialCache.java:
```diff
@@ -0,0 +1,71 @@
+
+    private final Map<String, Cache<? extends Object>> cacheMap = new HashMap<>();
+
+    public <C> Cache<C> createCache(String mechanism, Class<C> credentialClass) {
```
It seems all callers use the ScramCredential class. Is there a need to make this generic?

ijuma replied:
The idea is to make this generic to avoid a hard dependency on Scram-specific code in generic classes like `SocketServer` and `ConfigHandler`. That way, it should be easy to extend the cache mechanism to work with other authentication mechanisms like delegation tokens.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java:
```diff
@@ -121,7 +127,11 @@ public void configure(TransportLayer transportLayer, PrincipalBuilder principalB
     private void createSaslServer(String mechanism) throws IOException {
         this.saslMechanism = mechanism;
-        callbackHandler = new SaslServerCallbackHandler(Configuration.getConfiguration(), kerberosNamer);
+        ScramMechanism scramMechanism = ScramMechanism.forMechanismName(mechanism);
```
Perhaps we could just call isScram() instead?

rajinisivaram replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramMessages.java:
```diff
@@ -0,0 +1,272 @@
+            return toMessage().getBytes(StandardCharsets.UTF_8);
+        }
+
+        public String toMessage(byte[] messageBytes) {
```
Does this need to be a public method?

rajinisivaram replied:
Changed to protected.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClient.java:
```diff
@@ -0,0 +1,236 @@
+    private byte[] saltedPassword;
+    private ScramMessages.ClientFirstMessage clientFirstMessage;
+    private ScramMessages.ServerFirstMessage serverFirstMessage;
+    private ScramMessages.ClientFinalMessage clientFinalMessage;
```
Do this need to be an instance level var?

rajinisivaram replied:
It is needed to validate server final message which arrives after the response is sent, hence the value is stored.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClient.java:
```diff
@@ -0,0 +1,236 @@
+    private ScramMessages.ServerFirstMessage serverFirstMessage;
+    private ScramMessages.ClientFinalMessage clientFinalMessage;
+
+    public ScramSaslClient(ScramMechanism mechanism, String authorizationId, CallbackHandler cbh) throws NoSuchAlgorithmException {
```
authorizationId doesn't seem to be used?

rajinisivaram replied:
Removed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClient.java:
```diff
@@ -0,0 +1,236 @@
+                return clientFirstMessage.toBytes();
+
+            case RECEIVE_SERVER_FIRST_MESSAGE:
+                this.serverFirstMessage = new ServerFirstMessage(challenge);
```
Calls like this could generate SaslException. Should we catch those, set the state to fail and rethrow?

rajinisivaram replied:
Added a try-catch for SaslException. State is now set to `FAILED` only in the catch block since all failures get there.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClient.java:
```diff
@@ -0,0 +1,236 @@
+            case RECEIVE_SERVER_FINAL_MESSAGE:
+                ServerFinalMessage serverFinalMessage = new ServerFinalMessage(challenge);
+                if (serverFinalMessage.error() != null)
+                    throw new SaslException("Sasl authentication using " + mechanism + " failed with error: " + serverFinalMessage.error());
```
Should we set the state to FAILED before throwing the exception?

rajinisivaram replied:
Faled state now set in catch block before rethrowing SaslException.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java:
```diff
@@ -0,0 +1,212 @@
+                        callbackHandler.handle(new Callback[]{nameCallback, credentialCallback});
+                        this.scramCredential = credentialCallback.scramCredential();
+                        if (scramCredential == null)
+                            throw new SaslException("Authentication failed: Invalid user credentials");
```
Should we set the state to FAILED before throwing the exception?

rajinisivaram replied:
Similar to client, state is set before rethrowing SaslException.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java:
```diff
@@ -0,0 +1,212 @@
+        if (!isComplete())
+            throw new IllegalStateException("Authentication exchange has not completed");
+        String authzId = clientFirstMessage.authorizationId();
+        return authzId == null || authzId.length() == 0 ? username : authzId;
```
It seems that the client always leaves authorizationId empty in the first message. So, we could just return username here?

rajinisivaram replied:
Since spec says that authzid is used if set, it may be useful to do the check for interoperability with other SCRAM implementations.

junrao replied:
Ok, we can leave this as it is then.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java:
```diff
@@ -0,0 +1,212 @@
+            byte[] clientSignature = formatter.clientSignature(expectedStoredKey, clientFirstMessage, serverFirstMessage, clientFinalMessage);
+            byte[] computedStoredKey = formatter.storedKey(clientSignature, clientFinalMessage.proof());
+            if (!Arrays.equals(computedStoredKey, expectedStoredKey))
+                throw new SaslException("Invalid client credentials");
```
Should we set the state to FAILED here and in line 185 before throwing the exception?

rajinisivaram replied:
Done in try-catch before rethrowing exception.

junrao commented on clients/src/test/java/org/apache/kafka/common/security/scram/ScramFormatterTest.java:
```diff
@@ -0,0 +1,89 @@
+        for (String username : usernames) {
+            String saslName = formatter.saslName(username);
+            assertEquals(-1, saslName.indexOf(','));
+            assertEquals(-1, saslName.replace("=2C", "").replace("=3D", "").indexOf('='));
```
Could you add a comment on what this is testing?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -404,7 +407,7 @@ private[kafka] class Processor(val id: Int,
     metricTags,
     false,
-    ChannelBuilders.create(protocol, Mode.SERVER, LoginType.SERVER, channelConfigs, null, true))
+    ChannelBuilders.serverChannelBuilder(protocol, channelConfigs, credentialProvider.credentialCache))
```
With this change, the LoginType import is no longer needed.

rajinisivaram replied:
Removed.

junrao commented on core/src/main/scala/kafka/security/CredentialProvider.scala:
```diff
@@ -0,0 +1,56 @@
+
+import kafka.server.KafkaConfig
+
+import scala.collection.JavaConverters._
```
unused import here and above.

rajinisivaram replied:
Removed.

junrao commented on core/src/main/scala/kafka/security/CredentialProvider.scala:
```diff
@@ -0,0 +1,56 @@
+object CredentialProvider {
+  def userCredentialConfigs: ConfigDef = {
+    ScramMechanism.values.foldLeft(new ConfigDef) {
+      (c, m) => c.define(m.mechanismName.toLowerCase(Locale.ROOT), Type.STRING, null, Importance.MEDIUM, s"User credentials for SCRAM mechanism ${m.mechanismName}")
```
It seems that the dynamic config tool will take lower case mechanism name. What about the client/server config? It doesn't seem there is code to convert the lower case mechanism name to the ScramMechanism mechanism name, which is upper case?

rajinisivaram replied:
Lower-case mechanism names are used as the property name in ZK for consistency with other properties like quotas which are lower-case. Client/Server config specify SASL mechanism names, which have to be upper case (http://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xhtml). If we want to be consistent between ZK property name and mechanism name, perhaps we should just use upper case everywhere?

junrao replied:
Yes, maybe using upper case everywhere is better.

rajinisivaram replied:
@junrao Thank you, I have updated the PR and the examples in the KIP.

junrao reviewed (commented):
@rajinisivaram : Thanks for the PR. A few comments below. Also, do you plan to do the doc change in a separate PR?

rajinisivaram committed:
KAFKA-3751: Address review comments, remove SHA-224, SHA-384

rajinisivaram commented:
@junrao Thank you for the review. I have addressed most of the issues and left responses to review for the others. Also removed SHA-224 and SHA-384 and rebased. Have opened JIRAs for docs and system tests (KAFKA-4589, KAFKA-4590).

rajinisivaram committed:
KAFKA-3751: Use uppercase for scram property names

junrao commented on clients/src/main/java/org/apache/kafka/common/security/scram/ScramMessages.java:
```diff
@@ -0,0 +1,272 @@
+                throw new SaslException("Invalid SCRAM server first message format: " + message);
+            try {
+                this.iterations = Integer.parseInt(matcher.group("iterations"));
+                if (this.iterations <= 0)
```
It seems that checks like this is only done when the config change is propagated to the broker? Ideally, we want to fail the ConfigChangeCommand if the input is not valid.

rajinisivaram replied:
@junrao Yes, thank you. I have added validation of iterations in config command as well as during authentication. I think we should enforce a good minimum value. I have used 4096 for both mechanisms. For SCRAM-SHA-256, https://tools.ietf.org/html/rfc7677 defines 4096 as the minimum iterations. For SCRAM-SHA-512, there is no RFC yet, so using the same value.

rajinisivaram committed:
KAFKA-3751: Validate iteration count

junrao commented on clients/src/test/java/org/apache/kafka/common/security/scram/ScramFormatterTest.java:
```diff
@@ -0,0 +1,135 @@
+     */
+    @Test
+    public void rfc5802Example() throws Exception {
+        ScramFormatter formatter = new ScramFormatter("SHA-1", "HmacSHA1");
```
Since we didn't implement SHA-1, do we still need this test?

rajinisivaram replied:
@junrao Thank you for the reviews. No, this test is no longer required, removed.

junrao reviewed (commented):
Thanks for the patch. LGTM. Just one minor comment.

rajinisivaram committed:
KAFKA-3751: Removed unnecessary test

junrao commented:
@rajinisivaram : Thanks for the patch. LGTM
